### PR TITLE
fix: decode string with specified buffer size

### LINKED ISSUE #17: napi_create_string_utf8 should preserve null terminator
Opening for tracking purposes, original description of the issue here: https://github.com/toyobayashi/emnapi/pull/13#issuecomment-1372418731

### PR DESCRIPTION
**Pull request by toyobayashi**
Fixes #17

### CODE REVIEW

**Comment by RReverser** [comment]
This won't work correctly with pthreads, because `TextDecoder` works only on unshared memory. Note how Emscripten has to work around that with custom helper that switches to `.slice(...)` instead of `.subarray(...)` when memory is shared: https://github.com/emscripten-core/emscripten/blob/0d7014e1b20ead9739f1fce54b316152413601c3/src/runtime_strings.js#L115

File: packages/emnapi/src/error.ts [hunk 1]
```diff
@@ -201,7 +201,11 @@ function napi_fatal_error (location: const_char_p, location_len: size_t, message
   $from64('message')
   $from64('message_len')
 
-  abort('FATAL ERROR: ' + (location_len === -1 ? UTF8ToString(location) : UTF8ToString(location, location_len)) + ' ' + (message_len === -1 ? UTF8ToString(message) : UTF8ToString(message, message_len)))
+  abort('FATAL ERROR: ' +
+    (location_len === -1 ? UTF8ToString(location) : emnapiRt.utf8Decoder.decode(HEAPU8.subarray(location, location + location_len))) +
+    ' ' +
+    (message_len === -1 ? UTF8ToString(message) : emnapiRt.utf8Decoder.decode(HEAPU8.subarray(message, message + message_len)))
```
Comment (RReverser) [inline packages/emnapi/src/error.ts:207]: Maybe better to extract this into a helper `emnapiRt.readString`? Could be a bit cleaner than duplicating `=== -1 ? ...` everywhere.

File: packages/runtime/src/util.ts [hunk 2]
```diff
@@ -102,3 +102,80 @@ export const construct = supportReflect
     }
     return instance
   }
+
+export const utf8Decoder: { decode: (input: BufferSource) => string } = typeof TextDecoder === 'function'
+  ? new TextDecoder()
+  : {
+      decode (input: BufferSource) {
```
Comment (RReverser) [inline packages/runtime/src/util.ts:109]: I'd suggest to either exclude the fallback, or at least guard it under `#if TEXTDECODER == 2` etc like Emscripten does.

It's quite a bit of code for something that none of the modern engines needs.

File: packages/runtime/src/util.ts [hunk 3]
```diff
@@ -102,3 +102,80 @@ export const construct = supportReflect
...
+  ? new TextDecoder('utf-16le')
+  : {
+      decode (input: BufferSource) {
+        const isArrayBuffer = input instanceof ArrayBuffer
+        const isView = ArrayBuffer.isView(input)
+        if (!isArrayBuffer && !isView) {
+          throw new TypeError('The "input" argument must be an instance of ArrayBuffer or ArrayBufferView')
+        }
+        const bytes = isArrayBuffer ? new Uint16Array(input) : new Uint16Array(input.buffer, input.byteOffset, input.byteLength / 2)
+        return String.fromCharCode.apply(String, bytes as any)
```
Comment (RReverser) [inline packages/runtime/src/util.ts:179]: This is often tempting, but will fail with stack overflow on any large strings (maybe worth adding a test for that). It's important to use a loop here instead.

**Comment by toyobayashi** [comment]
Thanks. working in progress

**Comment by toyobayashi** [comment]
Done. https://github.com/toyobayashi/emnapi/commit/c8f11750037f1cf87d477920b9533db9e9235631

**Pull request merged**